planetiler-custommap/src/main/java/com/onthegomap/planetiler/custommap/ConfiguredFeature.java

```diff
@@ -0,0 +1,317 @@
+
+    Object constVal = attribute.constantValue();
+    if (constVal != null) {
+      return sf -> constVal;
+    }
+
+    String tagVal = attribute.tagValue();
+    if (tagVal != null) {
+      return tagValueProducer.getValueProducer(tagVal);
+    }
+
+    //Default to producing a tag identical to the input
+    return tagValueProducer.getValueProducer(attribute.key());
+  }
+
+  /**
+   * Generate logic which determines the minimum zoom level for a feature based on a configured pixel size limit.
+   * 
+   * @param minTilePercent - minimum percentage of a tile that a feature must cover to be shown
+   * @param minZoom        - global minimum zoom for this feature
+   * @param minZoomByValue - map of tag values to zoom level
+   * @return minimum zoom function
+   */
+  private static BiFunction<SourceFeature, Object, Byte> attributeZoomThreshold(Double minTilePercent, byte minZoom,
+    Map<Object, Byte> minZoomByValue) {
+
+    if (minZoom == 0 && minZoomByValue.isEmpty()) {
+      return null;
+    }
+
+    Function<SourceFeature, Byte> staticZooms =
+      sf -> (byte) Math.max(minZoom,
+        minZoomFromTilePercent(sf, minTilePercent));
+
+    if (minZoomByValue.isEmpty()) {
+      return (sf, key) -> staticZooms.apply(sf);
+    }
+
+    //Attribute value-specific zooms override static zooms
+    return (sourceFeature, key) -> minZoomByValue.getOrDefault(key, staticZooms.apply(sourceFeature));
+  }
+
+  private static byte minZoomFromTilePercent(SourceFeature sf, Double minTilePercent) {
+    if (minTilePercent == null) {
+      return 0;
+    }
+    try {
+      return (byte) (Math.log(minTilePercent / sf.area()) / LOG4);
+    } catch (GeometryException e) {
+      return 14;
+    }
+  }
+
+  /**
+   * Generates a function which produces a fully-configured attribute for a feature.
+   * 
+   * @param attribute - configuration for this attribute
+   * @return processing logic
+   */
+  private BiConsumer<SourceFeature, Feature> attributeProcessor(AttributeDefinition attribute) {
+    var tagKey = attribute.key();
+
+    var attributeMinZoom = attribute.minZoom();
+    attributeMinZoom = attributeMinZoom == null ? 0 : attributeMinZoom.byteValue();
+
+    var minZoomByValue = attribute.minZoomByValue();
+    minZoomByValue = minZoomByValue == null ? Map.of() : minZoomByValue;
+
+    //Workaround because numeric keys are mapped as String
+    minZoomByValue = tagValueProducer.remapKeysByType(tagKey, minZoomByValue);
+
+    var attributeValueProducer = attributeValueProducer(attribute);
+
+    var attrIncludeWhen = attribute.includeWhen();
+    var attrExcludeWhen = attribute.excludeWhen();
+
+    var attributeTest =
+      Expression.and(
+        attrIncludeWhen == null ? Expression.TRUE : matcher(attrIncludeWhen, tagValueProducer),
+        attrExcludeWhen == null ? Expression.TRUE : Expression.not(matcher(attrExcludeWhen, tagValueProducer))
+      );
```

Comment:
call simplify in the preprocessing step to avoid some unnecessary work at processing time